juliet/src/varint.rs

```diff
@@ -0,0 +1,315 @@
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct Varint32([u8; 6]);
+
+impl Debug for Varint32 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            v if v.is_sentinel() => f.write_str("Varint32::SENTINEL"),
+            _ => f.debug_tuple("Varint32").field(&self.0).finish(),
+        }
+    }
+}
+
+impl Varint32 {
+    /// `Varint32` sentinel.
+    ///
+    /// This value will never be parsed or generated by any encoded `u32`. It allows using a
+    /// `Varint32` as an inlined `Option<Varint32>`. The return value of `Varint32::len()` of the
+    /// `SENTINEL` is guaranteed to be `0`.
+    pub const SENTINEL: Varint32 = Varint32([0u8; 6]);
+
+    /// The maximum encoded length of a [`Varint32`].
+    pub const MAX_LEN: usize = 5;
+
+    /// Encodes a 32-bit integer to variable length.
+    #[inline]
+    pub const fn encode(mut value: u32) -> Self {
+        let mut output = [0u8; 6];
+        let mut count = 0;
+
+        while value > 0 {
+            output[count] = value as u8 & VARINT_MASK;
+            value >>= 7;
+            if value > 0 {
+                output[count] |= !VARINT_MASK;
+                count += 1;
+            }
+        }
+
+        output[5] = count as u8 + 1;
+        Varint32(output)
+    }
+
+    /// Returns the number of bytes in the encoded varint.
+    #[inline]
+    #[allow(clippy::len_without_is_empty)]
+    pub const fn len(self) -> usize {
+        self.0[5] as usize
+    }
+
+    /// Returns whether or not the given value is the sentinel value.
+    #[inline]
+    pub const fn is_sentinel(self) -> bool {
+        self.len() == 0
+    }
+
+    /// Decodes the contained `Varint32`.
+    ///
+    /// Should only be used in debug assertions, as `Varint32`s not meant to encoded/decoded cheaply
+    /// throughout their lifecycle. The sentinel value is decoded as 0.
+    pub(crate) const fn decode(self) -> u32 {
+        // Note: It is not possible to decorate this function with `#[cfg(debug_assertions)]`, since
+        //       `debug_assert!` will not remove the assertion from the code, but put it behind an
+        //       `if false { .. }` instead. Furthermore we also don't panic at runtime, as adding
+        //       a panic that only occurs in `--release` builds is arguably worse than this function
+        //       being called.
+
+        if self.is_sentinel() {
+            return 0;
+        }
+
+        match decode_varint32(self.0.as_slice()) {
+            Incomplete(_) | Fatal(_) => 0, // actually unreachable.
+            Success(v) => v.value,
+        }
+    }
+
+    /// Returns the length of the given value encoded as a `Varint32`.
+    #[inline]
+    pub const fn length_of(value: u32) -> usize {
+        if value < 128 {
+            return 1;
+        }
+
+        if value < 16384 {
+            return 2;
+        }
+
+        if value < 2097152 {
+            return 3;
+        }
+
+        if value < 268435456 {
```

Comment:
Or better yet (more readable IMO) `2<<N` syntax i.e. 2<<10, 2<<20, 2<<28 etc.

Reply:
Don't you mean `1<<N`? At least `1<<7` makes more sense to me here, since we're using 7 bits to encode the number.

Changed that way in `9cee0fef2`.

Reply:
> values in hex would be less cryptic :)

Well, in that case I'd like to present exhibit A:

> I think it might be worth adding a UT for the values at the boundary, such as:

```
        assert_eq!(Varint32::length_of(268435455), 4);
        assert_eq!(Varint32::length_of(268435456), 5);
```

`268435455` is `0xfffffff`, while `268435456` is `10000000`. Which is what `decode_known_values_and_crossover_points` tests :)